//  @file   config.proto
//  @brief  configuration for factorization machine
package dmlc.fm;

message Config {
  //
  // data
  //
  optional string train_data = 1;
  optional string val_data = 2;
  optional string test_data = 3;
  optional string data_format = 4 [default = "libsvm"];

  optional bool use_worker_local_data = 6 [default = false];

  // partition a file into npart
  optional int32 num_parts_per_file = 5 [default = 1];

  //
  // model
  //
  optional string model_in = 10;
  // load model from iter = *load_model*
  optional int32 load_model = 11;

  optional string model_out = 12;
  // save model for every *save_model* data pass
  optional int32 save_model = 13;

  // for w
  optional float lambda_l2 = 20;
  optional float lambda_l1 = 21;
  // learning rate, often in the format lr_eta / (lr_beta + x), where x is
  // user-defined, such as sqrt(iter), or the cumulative gradient on adagrad
  optional float lr_eta = 41;
  optional float lr_beta = 42;

  enum Task {
    TRAIN = 1;
    PREDICT = 2;
  }
  optional Task task = 15 [default = TRAIN];

  // for V
  message Embedding {
    optional int32 dim = 1;
    optional int32 threshold = 2;

    optional float lambda_l2 = 3;
    // learning rate
    optional float lr_eta = 4 [default = .01];
    optional float lr_beta = 5 [default = 1];
    // uniformly random weight in [-init_scale, +init_scale]
    optional float init_scale = 6 [default = .01];

    // experimetal
    optional float dropout = 7;
    optional float grad_clipping = 8;

    optional float grad_normalization = 9 [default = 0];
  }

  repeated Embedding embedding = 22;

  //
  // optimization methods
  //

  // the size of minibatch for online methods
  optional int32 minibatch = 31 [default = 1000];

  // the maximal number of data passes
  optional int32 max_data_pass = 32 [default = 10];

  // print the progress every n sec
  optional float disp_itv = 33 [default = 1];

  // stop if the objective values exceeds *max_objv* (often used in parameter
  // tuning)
  optional float max_objv = 34;

  //
  // system performance
  //

  // maximal allowed delay during synchronization
  optional int32 max_delay = 50 [default = 0];
  // communication filters
  optional bool key_cache = 51 [default = true];
  optional bool msg_compression = 52 [default = true];
  // 0 means no fixing
  optional int32 fixed_bytes = 53 [default = 1];

  // number of threads used to process one minibatch
  optional int32 num_threads = 60 [default = 2];

  /// exmperimental
  optional bool l1_shrk = 100 [default = true];
  optional int32 rand_shuffle = 101 [default = 10];

  optional bool early_stop = 102 [default = false];
  optional float min_objv_decr = 103 [default = .00001];

  optional bool grad_normalization = 105 [default = false];
}
